ReactNativeClient/lib/models/Tag.js

```diff
@@ -174,23 +342,131 @@ class Tag extends BaseItem {
+	}
+
+	static async moveTag(tagId, parentTagId) {
+		if (tagId === parentTagId
+				|| (await Tag.descendantTagIds(tagId)).includes(parentTagId)) {
+			throw new Error(_('Cannot move tag to this location.'));
+		}
+		if (!parentTagId) parentTagId = '';
+
+		let tag = await Tag.load(tagId);
+		if (!tag) return;
+
+		const oldParentTagId = tag.parent_id;
+
+		// Save new parent id
+		tag.parent_id = parentTagId;
+		tag = await Tag.save(tag, { userSideValidation: true });
+
+		if (parentTagId !== oldParentTagId) {
+			// If the parent tag has changed, and the ancestor doesn't
+			// have notes attached, then remove it
+			const oldParentWithCount = await Tag.loadWithCount(oldParentTagId);
+			if (!oldParentWithCount) {
+				await Tag.delete(oldParentTagId, { deleteChildren: false, deleteNotelessParents: true });
+			}
+		}
+
+		return tag;
+	}
+
+	static async renameNested(tag, newTitle) {
+		const oldParentId = tag.parent_id;
+
+		tag = await Tag.saveNested(tag, newTitle, { fields: ['title', 'parent_id'], userSideValidation: true });
+
+		if (oldParentId !== tag.parent_id) {
+			// If the parent tag has changed, and the ancestor doesn't
+			// have notes attached, then remove it
+			const oldParentWithCount = await Tag.loadWithCount(oldParentId);
+			if (!oldParentWithCount) {
+				await Tag.delete(oldParentId, { deleteChildren: false, deleteNotelessParents: true });
+			}
+		}
+		return tag;
+	}
+
+	static async saveNested(tag, fullTitle, options) {
+		if (!options) options = {};
+		// The following option is used to prevent loops in the tag hierarchy
+		if (!('mainTagId' in options) && tag.id) options.mainTagId = tag.id;
+
+		if (fullTitle.startsWith('/') || fullTitle.endsWith('/')) {
+			throw new Error(_('Tag name cannot start or end with a `/`.'));
+		} else if (fullTitle.includes('//')) {
+			throw new Error(_('Tag name cannot contain `//`.'));
+		}
+
+		let parentId = '';
+		// Check if the tag is nested using `/` as separator
+		const separator = '/';
+		const i = fullTitle.lastIndexOf(separator);
+		if (i !== -1) {
+			const parentTitle = fullTitle.slice(0,i);
+			tag.title = fullTitle.slice(i + 1);
+
+			// Try to get the parent tag
+			const parentTag = await Tag.loadByTitle(parentTitle);
+			// The second part of the conditions ensures that we do not create a loop
+			// in the tag hierarchy
+			if (parentTag &&
+				!('mainTagId' in options
+					&& (options.mainTagId === parentTag.id
+						|| (await Tag.descendantTagIds(options.mainTagId)).includes(parentTag.id)))
+			) {
+				parentId = parentTag.id;
+			} else {
+				// Create the parent tag if it doesn't exist
+				const parentOpts = {};
+				if ('mainTagId' in options) parentOpts.mainTagId = options.mainTagId;
+				const parentTag = await Tag.saveNested({}, parentTitle, parentOpts);
+				parentId = parentTag.id;
+			}
+		} else {
+			// Tag is not nested so set the title to full title
+			tag.title = fullTitle;
+		}
+
+		// Set parent_id
+		tag.parent_id = parentId;
```

Comment:
Please save the tag first to a new object to avoid mutations - `const newTag = Object.assign({}, tag)`